src/compiler/crystal/semantic/exhaustiveness_checker.cr

```diff
@@ -0,0 +1,644 @@
+  # For example, the BoolTarget will have one subtarget for the value `false` and
+  # one for the value `true`. When a pattern like `{false, ...}` is passed to it,
+  # only patterns for the `false` subtarget will be covered.
+  abstract class Target
+    # The type this target is based on.
+    getter type
+
+    def initialize(@type : Type)
+      @type_covered = false
+    end
+
+    # Tries to cover this target with the given pattern.
+    # By default, a TypePatteren will cover a target.
+    # Other, more specific, patterns will partially cover a target.
+    def cover(pattern : Pattern) : Nil
+      case pattern
+      when TypePattern
+        if @type.implements?(pattern.type)
+          @type_covered = true
+        end
+      when UnderscorePattern
+        @type_covered = true
+      when EnumMemberPattern, EnumMemberNamePattern, BoolPattern
+        # No cover
+      end
+    end
+
+    # Covers this target and subsequent subtargets with the patterns starting
+    # at index.
+    abstract def cover(patterns : Array(Pattern), index : Int32) : Nil
+
+    # Removes covered subtargets from this target, and returns whether
+    # this target ended up being entirely covered.
+    abstract def reject_covered! : Bool
+
+    # Was this target covered?
+    abstract def covered? : Bool
+
+    # What are the cases that we didn't cover?
+    abstract def missing_cases : Array(String)
+
+    # Add subtargets for the given type groups, starting at index.
+    abstract def add_subtargets(type_groups : Array(Array(Type)), index : Int32) : Nil
+  end
+
+  # A bool target. Subtargets are the `false` and `true` literals.
+  class BoolTarget < Target
+    property? found_true = false
+    property? found_false = false
+    property! subtargets : Hash(Bool, Array(Target))
+
+    def cover(pattern : Pattern) : Nil
+      super
+
+      case pattern
+      when BoolPattern
+        if pattern.value
+          @found_true = true
+        else
+          @found_false = true
+        end
+      when UnderscorePattern
+        @found_true = true
+        @found_false = true
+      when TypePattern, EnumMemberPattern, EnumMemberNamePattern
+        # No cover
+      end
+    end
+
+    def cover(patterns : Array(Pattern), index : Int32) : Nil
+      if index == patterns.size - 1
+        cover(patterns.last)
+        return
+      end
+
+      pattern = patterns[index]
+      case pattern
+      when TypePattern
+        if @type.implements?(pattern.type)
+          subtargets.each_value &.each &.cover(patterns, index + 1)
+        end
+      when BoolPattern
+        subtargets[pattern.value].each &.cover(patterns, index + 1)
+      when UnderscorePattern
+        subtargets.each do |key, targets|
+          subtargets.each_value &.each &.cover(patterns, index + 1)
+        end
+      when EnumMemberPattern, EnumMemberNamePattern
+        # No cover
+      end
+    end
+
+    def reject_covered! : Bool
+      if subtargets = @subtargets
```

Comment:
💭  These `if subtargets = @subtargets` are everywhere.

I thought about having `Target` and `CompositeTarget` to distinguish the case of a target holding subtargets... but I think the way it's done right now works just as well. Plus, adding a Target means you have to consider it both for the regular case and the tuple case. The code is a bit more stateful, but it works fine.

We can always refactor this, if needed, later on without causing any breaking change.